### ISSUE TITLE
Communication protocol

### ISSUE THREAD

**Issue by sjrct** [iiag/iiag#1]
Reflective of issue [#43](https://github.com/iiag/iiag-legacy/issues/43) in iiag-legacy. We need a communication protocol between the server and the client.

Perhaps event based? So when an event happens that the client can see, an event message is sent to the client, and they can update accordingly. Perhaps there should be a method for the client to query for other information as well, though I think the events should in theory provide sufficient information provided the client and server do not go out of sync.

**Comment by Lundar** [comment]
What are peoples opinions on what networking library we use? I would prefer to avoid BSD sockets as they are not the most portable option we could have. [ENET](http://enet.bespin.org/) is good performance wise, but has the drawback of being needed at both ends, and limiting client languages. Some simple wrapper for sockets would be ideal. [nanmsg](http://nanomsg.org/), or [libuv](https://github.com/libuv/libuv) might work. other suggestions?

**Comment by Grissess** [comment]
AFAIK Berkeley sockets _are_ rather portable (cf. WinSock). The problem may be that they're too cumbersome or low-level :P. 

Frankly, though, I'm most comfortable at that level--with some Python niceties, of course.

**Comment by sjrct** [comment]
They only concern I have in any respect of this matter is limiting our options. WinSock is a little different from the typical BSD sockets, but nothing that can't be `#ifdef`ed around, so it is a fine option in my mind. With regard to that I would prefer to steer clear of ENET, but nanomsg looks pretty good in that respect. libuv would probably also be fine as it provides direct access to TCP/UDP sockets which should be supported by virtually every networkable language out there, though it is perhaps somewhat node.js oriented.

In summary, I am fine with any option so long as it is portable.

**Comment by Grissess** [comment]
As mentioned before, I _like_ sockets, and I'd gladly write socket-level code.

However, nanomsg up there looks pretty nice--and has a lot of language bindings (including Python!). For our purposes, I don't know how much using it would win us over the regular socket API, though.

For absolutely stupid purposes, I have implemented [more complicated protocols](https://github.com/Grissess/nom/blob/master/packet.py) in more dynamic languages--these especially have the nice property of being forward-compatible, at the cost of a little complexity to implement (a mapping type and arbitrary reference cycle support are required).

...but I think the worse-is-better approach is just fine here :P

**Comment by sjrct** [comment]
If the need becomes clear that we need something more robust, then we use that more robust thing at that point in time, right now I don't think it will be that big a deal either way. Are BSD sockets fine with everyone for now? If the winsock difference become to great or there is need for more complex routing we can use something else then. Sensible?

**Comment by Grissess** [comment]
I'm fine with this, but I was the one to propose it, so I'd seek other opinions :P

**Comment by Lundar** [comment]
BSD sockets will work. Its not the best, but it may be a second best. I'm not sure what the best library is.
